<i>Copyright (c) Recommenders contributors.</i>

<i>Licensed under the MIT License.</i>

# Item2Item recommendations with DKN 
The second task is about knowledge-aware item-to-item recommendations. We still use DKN for demonstration. 
The learning framework is illustrated as follows:
<img src="https://raw.githubusercontent.com/recommenders-team/resources/main/Item2item-framework.JPG"  width="500">

In [1]:
import os
import time

from recommenders.models.deeprec.models.pytorch.dkn_item2item import DKNItem2Item
from recommenders.utils.notebook_utils import store_metadata

In [2]:
data_path = 'data_folder/my/DKN-training-folder'

EPOCHS = 10
BATCH_SIZE = 32
LEARNING_RATE = 0.0002
RANDOM_SEED = 42  # Set this to None for non-deterministic result

In [3]:
train_file = os.path.join(data_path, r'item2item_train_instances.txt')
valid_file = os.path.join(data_path, r'item2item_valid_instances.txt')
news_feature_file = os.path.join(data_path, r'../paper_feature.txt')
word_embedding_file = os.path.join(data_path, r'word_embedding.npy')
entity_embedding_file = os.path.join(data_path, r'entity_embedding.npy')
context_embedding_file = os.path.join(data_path, r'context_embedding.npy')
infer_embedding_file = os.path.join(data_path, r'infer_embedding_item2item.txt')
model_dir = os.path.join(data_path, 'save_models')

To build an item2item recommendation model based on the Recommenders repo, only two pieces change compared with DKN:
1. Data loader: `DKNItem2ItemDataset` in `recommenders/models/deeprec/io/dkn_dataset.py`
2. Model: `DKNItem2Item` in `recommenders/models/deeprec/models/pytorch/dkn_item2item.py`

Both reuse the knowledge-aware CNN and the news feature loading of DKN.

A special parameter is `neg_num`. It indicates how many negative instances exist in a group for softmax computation.
Training and validation instances are organized as follows:

<img src="https://raw.githubusercontent.com/recommenders-team/resources/main/item2item-instances.JPG" width="700">

In [4]:
pairwise_metrics = ['group_auc', 'mean_mrr', 'ndcg@2;4;6']

In [5]:
model = DKNItem2Item(
    news_feature_file=news_feature_file,
    word_embedding_file=word_embedding_file,
    neg_num=9,
    entity_embedding_file=entity_embedding_file,
    context_embedding_file=context_embedding_file,
    filter_sizes=[1, 2, 3],
    num_filters=50,
    init_method='uniform',
    init_value=0.01,
    seed=RANDOM_SEED,
)

In [6]:
t01 = time.time()
print(model.run_eval(valid_file, batch_size=BATCH_SIZE, pairwise_metrics=pairwise_metrics))
t02 = time.time()
print((t02 - t01) / 60)

{'group_auc': 0.8347, 'mean_mrr': 0.6729, 'ndcg@2': 0.6198, 'ndcg@4': 0.6947, 'ndcg@6': 0.7264}
2.4153950850168866


In [7]:
model = model.fit(
    train_file,
    valid_file,
    epochs=EPOCHS,
    batch_size=BATCH_SIZE,
    learning_rate=LEARNING_RATE,
    embed_l2=0.0,
    layer_l2=0.0,
    max_grad_norm=0.5,
    pairwise_metrics=pairwise_metrics,
    model_dir=model_dir,
    save_epoch=1,
)

at epoch 1
train info: loss:51.14471993375305
eval info: group_auc:0.9462, mean_mrr:0.8636, ndcg@2:0.8556, ndcg@4:0.8847, ndcg@6:0.8914


at epoch 2
train info: loss:48.820340668914824
eval info: group_auc:0.9489, mean_mrr:0.875, ndcg@2:0.8696, ndcg@4:0.8932, ndcg@6:0.8993


at epoch 3
train info: loss:47.732691279385406
eval info: group_auc:0.95, mean_mrr:0.8815, ndcg@2:0.8762, ndcg@4:0.8976, ndcg@6:0.9036


at epoch 4
train info: loss:46.984521386082186
eval info: group_auc:0.9499, mean_mrr:0.885, ndcg@2:0.8799, ndcg@4:0.8999, ndcg@6:0.9057


at epoch 5
train info: loss:46.45142303954918
eval info: group_auc:0.9495, mean_mrr:0.8856, ndcg@2:0.8804, ndcg@4:0.9, ndcg@6:0.9057


at epoch 6
train info: loss:46.10383621971405
eval info: group_auc:0.9502, mean_mrr:0.8861, ndcg@2:0.8817, ndcg@4:0.901, ndcg@6:0.9062


at epoch 7
train info: loss:45.85681621508051
eval info: group_auc:0.949, mean_mrr:0.8864, ndcg@2:0.8812, ndcg@4:0.9008, ndcg@6:0.9059


at epoch 8
train info: loss:45.68403733175747
eval info: group_auc:0.951, mean_mrr:0.8894, ndcg@2:0.8845, ndcg@4:0.9038, ndcg@6:0.9087


at epoch 9
train info: loss:45.52055471522259
eval info: group_auc:0.9493, mean_mrr:0.8878, ndcg@2:0.8831, ndcg@4:0.9019, ndcg@6:0.9068


at epoch 10
train info: loss:45.40995791900795
eval info: group_auc:0.9505, mean_mrr:0.8889, ndcg@2:0.8844, ndcg@4:0.9029, ndcg@6:0.9081


In [8]:
res = model.run_eval(valid_file, batch_size=BATCH_SIZE, pairwise_metrics=pairwise_metrics)
print(res)

{'group_auc': 0.9505, 'mean_mrr': 0.8889, 'ndcg@2': 0.8844, 'ndcg@4': 0.9029, 'ndcg@6': 0.9081}


In [9]:
# Record results for tests - ignore this cell
for name, value in res.items():
    store_metadata(name, value)

In [10]:
model = model.run_get_embedding(news_feature_file, infer_embedding_file, batch_size=BATCH_SIZE)

Again, we compre with DKN performance between using knowledge entities or without using knowledge entities (DKN(-)):

| Models | Group-AUC | MRR |NDCG@2 | NDCG@4 |
| :------| :------: | :------: | :------: | :------ |
| DKN | 0.9557 | 0.8993 | 0.8951 | 0.9123 |
| DKN(-) | 0.9506 | 0.8817 | 0.8758 | 0.8982 |

